### PR TITLE
Messages

### PR DESCRIPTION
**Pull request by TomAugspurger**

### CODE REVIEW

File: distributed/comm/ucx.py [hunk 1]
```diff
@@ -103,14 +102,55 @@ def peer_address(self):
...
+        # I've included it now for unpacking.
+        lengths = [nbytes(frame) for frame in frames]
+        length_bytes = ([struct.pack('Q', len(frames))] +
+                        [struct.pack('Q', x) for x in lengths])
+        b = b''.join(length_bytes + frames)  # small enough, send in one go
+
+        # TODO: exception handling
+        length_bytes = sys.getsizeof(b)
+        await self.ep.send_msg(b, length_bytes)
+        return length_bytes
```
Comment (mrocklin) [inline distributed/comm/ucx.py]: So you shouldn't need to do much of this.  Ideally it looks a bit more like the following:

```python
frames = await to_frames(...)
await self.ep.send_msg(struct.pack('Q', len(frames)))  # send number of frames
for frame in frames:
    await self.ep.send_msg(frame)
```

UCX will handle the other logic around the number of bytes in each frame for us.  It's a bit smarter than Tornado here.

**Review by mrocklin — COMMENTED** [review]
One comment.  Also, thoughts on rebasing this off of master rather than the ucx branch?

File: distributed/comm/ucx.py [hunk 2]
```diff
@@ -103,14 +102,55 @@ def peer_address(self):
...
+        msg = b''.join(frames)
+        # nbytes = sys.getsizeof(msg)
+
+        # Do we need any of the `stuck.pack` stuff? Is that just
+        # padding the messages for TCP?
+        # I've included it now for unpacking.
+        lengths = [nbytes(frame) for frame in frames]
+        length_bytes = ([struct.pack('Q', len(frames))] +
+                        [struct.pack('Q', x) for x in lengths])
+        b = b''.join(length_bytes + frames)  # small enough, send in one go
```
Comment (mrocklin) [inline distributed/comm/ucx.py]: Mostly I'm trying to avoid this memory copy and the coercion of memoryviews into bytes objects

**Comment by TomAugspurger** [comment]
Thanks, trying that out now.

> Also, thoughts on rebasing this off of master rather than the ucx branch?

I'm planning to get this into my ucx branch relatively quickly. And then I think I'll make a proper PR to distributed shortly after that (after some cleanup and writing additional tests). Does that sound OK?

**Comment by mrocklin** [comment]
Sounds good

**Comment by TomAugspurger** [comment]
Thanks, that's quite a bit simpler. with 250b573

Still failing on the issue with the connector stealing messages from the listener, debugging that next.

File: distributed/comm/tests/test_ucx.py [hunk 3]
```diff
@@ -56,8 +56,7 @@ def test_ucx_specific():
 
         async def handle_comm(comm):
             # XXX: failures here don't fail the build yet
-            import pdb; pdb.set_trace()
-            msg = yield comm.read()
+            msg = await comm.read()
```
Comment (TomAugspurger) [inline distributed/comm/tests/test_ucx.py:59]: I'm guessing this isn't the only time I make this mistake :) Look for strange GeneratorExits.

**Pull request merged**